### PR TITLE
introspect: enums

### PR DESCRIPTION
**Pull request by shramee**
### Enums and Introspect impls

```rs
#[derive(Serde, Copy, Drop, Introspect)]
enum PlainEnum {
    Left: (),
    Right: (),
}

impl PlainEnumSchemaIntrospection of dojo::database::schema::SchemaIntrospection<PlainEnum> {
    #[inline(always)]
    fn size() -> usize {
        1
    }

    #[inline(always)]
    fn layout(ref layout: Array<u8>) {
        layout.append(8);
    }

    #[inline(always)]
    fn ty() -> dojo::database::schema::Ty {
        dojo::database::schema::Ty::Enum(
            dojo::database::schema::Enum {
                name: 'Direction',
                attrs: array![].span(),
                children: array![
                    (
                        'Left',
                        serialize_member_type(@dojo::database::schema::Ty::Tuple(array![].span()))
                    ),
                    (
                        'Right',
                        serialize_member_type(@dojo::database::schema::Ty::Tuple(array![].span()))
                    )
                ]
                    .span()
            }
        )
    }
}



#[derive(Serde, Copy, Drop, Introspect)]
enum EnumPrimitive {
    Left: (u16,),
    Right: (u16,),
}

impl EnumPrimitiveSchemaIntrospection of dojo::database::schema::SchemaIntrospection<EnumPrimitive> {
    #[inline(always)]
    fn size() -> usize {
        2
    }

    #[inline(always)]
    fn layout(ref layout: Array<u8>) {
        layout.append(8);
        layout.append(16);
    }

    #[inline(always)]
    fn ty() -> dojo::database::schema::Ty {
        dojo::database::schema::Ty::Enum(
            dojo::database::schema::Enum {
                name: 'Direction',
                attrs: array![].span(),
                children: array![
                    (
                        'Left',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['u16'].span())
                        )
                    ),
                    (
                        'Right',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['u16'].span())
                        )
                    )
                ]
                    .span()
            }
        )
    }
}



#[derive(Serde, Copy, Drop, Introspect)]
enum EnumTuple {
    Left: (u8, u8),
    Right: (u8, u8),
}

impl EnumTupleSchemaIntrospection of dojo::database::schema::SchemaIntrospection<EnumTuple> {
    #[inline(always)]
    fn size() -> usize {
        3
    }

    #[inline(always)]
    fn layout(ref layout: Array<u8>) {
        layout.append(8);
        layout.append(8);
        layout.append(8);
    }

    #[inline(always)]
    fn ty() -> dojo::database::schema::Ty {
        dojo::database::schema::Ty::Enum(
            dojo::database::schema::Enum {
                name: 'Direction',
                attrs: array![].span(),
                children: array![
                    (
                        'Left',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['u8', 'u8'].span())
                        )
                    ),
                    (
                        'Right',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['u8', 'u8'].span())
                        )
                    )
                ]
                    .span()
            }
        )
    }
}



#[derive(Serde, Copy, Drop, Introspect)]
enum EnumCustom {
    Left: Vec2,
    Right: Vec2,
}

impl EnumCustomSchemaIntrospection of dojo::database::schema::SchemaIntrospection<EnumCustom> {
    #[inline(always)]
    fn size() -> usize {
        dojo::database::schema::SchemaIntrospection::<Vec2>::size() + 1
    }

    #[inline(always)]
    fn layout(ref layout: Array<u8>) {
        layout.append(8);
        dojo::database::schema::SchemaIntrospection::<Vec2>::layout(ref layout);
    }

    #[inline(always)]
    fn ty() -> dojo::database::schema::Ty {
        dojo::database::schema::Ty::Enum(
            dojo::database::schema::Enum {
                name: 'Direction',
                attrs: array![].span(),
                children: array![
                    (
                        'Left',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['Vec2'].span())
                        )
                    ),
                    (
                        'Right',
                        serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(array!['Vec2'].span())
                        )
                    )
                ]
                    .span()
            }
        )
    }
}
```
```

### CODE REVIEW

File: crates/dojo-lang/src/plugin_test_data/introspect [hunk 1]
```diff
@@ -82,6 +106,186 @@ impl Vec2SchemaIntrospection of dojo::database::schema::SchemaIntrospection<Vec2
...
+    fn ty() -> dojo::database::schema::Ty {
+        dojo::database::schema::Ty::Enum(
+            dojo::database::schema::Enum {
+                name: 'Direction',
+                attrs: array![].span(),
+                children: array![
+                    (
+                        'Left',
+                        serialize_member_type(
+                            @dojo::database::schema::Ty::Tuple(array!['Vec2'].span())
```
Comment (tarrencev) [inline crates/dojo-lang/src/plugin_test_data/introspect:271]: This should have the full Vec2 subtree

File: crates/dojo-lang/src/plugin_test_data/introspect [hunk 2]
```diff
@@ -82,6 +106,186 @@ impl Vec2SchemaIntrospection of dojo::database::schema::SchemaIntrospection<Vec2
...
+    fn ty() -> dojo::database::schema::Ty {
+        dojo::database::schema::Ty::Enum(
+            dojo::database::schema::Enum {
+                name: 'Direction',
+                attrs: array![].span(),
+                children: array![
+                    (
+                        'Left',
+                        serialize_member_type(
+                            @dojo::database::schema::Ty::Tuple(array!['u16'].span())
```
Comment (tarrencev) [inline crates/dojo-lang/src/plugin_test_data/introspect:178]: In this case, it should derive a Tuple of Ty::Primitives

File: crates/dojo-lang/src/plugin_test_data/introspect (hunk 1 above)
Comment (shramee) [reply crates/dojo-lang/src/plugin_test_data/introspect]: Now I know why it was nested span...

```rs
                    (
                        'Left',
                        dojo::database::schema::serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(
                                array![
                                    dojo::database::schema::serialize_member_type(
                                        @dojo::database::schema::Ty::Primitive('u8')
                                    ),
                                    dojo::database::schema::serialize_member_type(
                                        @dojo::database::schema::Ty::Primitive('u8')
                                    )
                                ]
                                    .span()
                            )
                        )
                    ),
```

File: crates/dojo-lang/src/plugin_test_data/introspect (hunk 2 above)
Comment (shramee) [reply crates/dojo-lang/src/plugin_test_data/introspect]: Now I know why it was nested span...

```rs
                    (
                        'Left',
                        dojo::database::schema::serialize_member_type(
                            @dojo::database::schema::Ty::Tuple(
                                array![
                                    dojo::database::schema::serialize_member_type(
                                        @dojo::database::schema::Ty::Primitive('u8')
                                    ),
                                    dojo::database::schema::serialize_member_type(
                                        @dojo::database::schema::Ty::Primitive('u8')
                                    )
                                ]
                                    .span()
                            )
                        )
                    ),
```

**Pull request merged**